### ISSUE TITLE
Ability to add local code to Sourcegraph

### ISSUE THREAD

**Issue by keegancsmith** [sourcegraph/sourcegraph#47373]

**Comment by keegancsmith** [comment]
We have Noah working on packages and Peter working on Perforce. Both of these are working in the layer around git repository syncing in "non-traditional" ways. Additionally I know @mrnugget cares about improving this area. So when we tackle this (soon? now?) we should avoid hacks that will come back to bite the Repo Management team in 3 months time.

My initial idea is to get gitserver to just use the clones that already exist rather than cloning them. This will require some sort of "read-only" mode in gitserver to avoid us messing with peoples code. I initially thought this could live in the part were we convert an `api.RepoName` into a `GitDir`. Thorsten pointed out a likely better idea than directly pointing, but instead setup a `VCSSyncer` which just creates symlinks.

Complications/unknowns of relying on user direct clones:
- Does batch changes mutate the repo at all? Is it done in a safe way (detached commits?)
- Won't contain all code on remote. EG will be missing refs / we won't know what the remote HEAD is (ie which is the main branch). For search this is actually nice because we search the code that is checked in at HEAD.
- User expectations around dirty working copies. IE we won't be searching the index or working copy, only things that appear in commits.
- How do we match up clones to an ExternalService such that batch changes/link to external/etc still work?

An alternative? Maybe clone like we normally do, but make it possible to use users local clones to avoid fetching everything.

Ruled out: Use things like GIT_ALTERNATES/etc. We can't rely on users clones to not remove objects.

Simple: make it work like src-serve-git. Easy to code, fits into the current arch. But you lose out on a lot (no code host specific connections) and we duplicate clones.

cc @sourcegraph/repo-management please if you have ideas on how we can accomplish this I would appreciate ideas. We are just thinking about this, so desperate for ideas to help us find a solution that is acceptable.

**Comment by peterguy** [comment]
do any of the concepts fund in [src-expose](https://github.com/sourcegraph/sourcegraph/tree/main/dev/src-expose) help?

**Comment by keegancsmith** [comment]
> do any of the concepts fund in src-expose help?

Yup, I think that will be a reasonable first step just to get something
working. But then there are thoughts around avoiding the need to clone,
integration into "native" external services so we get things like batch
changes / external links / etc. However, if I remember how we designed
src-expose we can make that API pretty much expose any field that we end
up storing in the `repo` table which means we could probably do a lot.

**Comment by mrnugget** [comment]
> However, if I remember how we designed src-expose we can make that API pretty much expose any field that we end up storing in the `repo` table which means we could probably do a lot.

Ah, that could be an interesting way of handling it. You expose the local file path via `src-expose`, it gets stord as metadata in `repo.metadata` and then on `gitserver` side we can special-case that and not do a full clone

**Comment by slimsag** [comment]
@keegancsmith Yes, agreed, and with respect to timeline it's gotta be sometime in the next 6 weeks :) If you want to start diving into this that sounds great to me, I will get to it later if not.

With respect to hackiness, I definitely prefer a proper implementation - though if we run into time constraints we could of course restrict logic to App.

> Won't contain all code on remote. EG will be missing refs / we won't know what the remote HEAD is 

This seems fine to me and reasonable (Sourcegraph only knows what code you have on your local system.)

> User expectations around dirty working copies. IE we won't be searching the index or working copy, only things that appear in commits.

I think this is really unfortunate and it'd be ideal if we could handle dirty working copies somehow. I definitely haven't thought of how, but just like it'd be weird for an editor to not show the on-FS state of your code, I think it'd be weird for Sourcegraph not to. I think most users would see this as a bug / serious drawback.

Not saying we have to address this for Starship MVP (6 weeks from now), and I'd also say batch changes on local repos is also not a strict requirement for the MVP-but I think both are necessary longer term, so would be good to think about now if possible.

**Comment by keegancsmith** [comment]
> @keegancsmith Yes, agreed, and with respect to timeline it's 
> gotta be sometime in the next 6 weeks :) If you want to start 
> diving into this that sounds great to me, I will get to it later 
> if not.

Cool, more than likely I'll get started on this next week (ie once 
merge is done and dusted)

>> User expectations around dirty working copies. IE we won't be 
>> searching the index or working copy, only things that appear in 
>> commits. 
> 
> I think this is really unfortunate and it'd be ideal if we could 
> handle dirty working copies somehow. I definitely haven't 
> thought of how, but just like it'd be weird for an editor to not 
> show the on-FS state of your code, I think it'd be weird for 
> Sourcegraph not to. I think most users would see this as a bug / 
> serious drawback.

I think designing for this will fundamentally change how we do 
this. The reason being is Sourcegraph is commit centric. I've had 
lots of offline conversations at merge, but haven't talked too 
much about this. I will have a few more now. Maybe the work on 
perforce native will allow us to handle this?

Either way, as an MVP we will be able to search local commits easily. I
think that is straightforward enough that I might just implement that.

**Comment by keegancsmith** [comment]
https://github.com/sourcegraph/sourcegraph/pull/47673 is a proof of concept that I think we can land in main. It does the simplest possible approach, using a copy of src serve-git to automatically import local repositories. There is no connection to the app-setup work yet, but I think this is a good start and possibly unblocks ideas on how we integrate it with that / what we really want from a features perspective / etc.

Until there is more feedback, I will start tackling other blockers to get ahead of those.

**Comment by keegancsmith** [comment]
@vovakulikov @gl-srgr I am responding to you here instead of slack because slack is not great for long discussion / we do not retain the conversations. [Original thread](https://sourcegraph.slack.com/archives/C04J6HSHVJ5/p1676494828758359)

> is it gonna be just yet another kind of code host and this mean that this would be part of externalServices API?

Right now it is implemented as the existing external service "Other". There are downsides to this as detailed above. The main being is we have code host connection specific behaviour like in batch changes / things like links to the code host.

I imagine we can overcome this, but it won't fit as neatly into the current design of the external service table. I think we need to be driven by use cases to help shape this.

> I and Gary Lee were discussing this earlier that it woul be nice to use existing types around external services, this would simplify setup wizard client a lot.

Yes, right now it is just an external service (with kind Other). The APIs should be able to be adjusted to give a preview of repos without actually syncing to DB. If that turns out to be difficult we can make it local specific. I expect it to fit neatly into the Syncer design though.

> Also if we had local repositories with externalService it would be possible to fetch info about progress, like how many repositories this path gave, what is status of resolved repositoris, &#x2026;etc

This is all possible right now using the normal syncer. IE we do discovery on the FS (pretty quick), these end up as entries in the repos table. Then gitserver\_repos table will update with status of cloning / etc. If discovery is slow, it is possible to adjust the API to be streaming.

This doesn't give you a way to preview, but we can probaly re-architect parts of repo-updater to give us "previews" for all external services relatively easily (if the interfaces haven't changed too much since I last looked at that code).

> I think we also need to have some kind of validation there, imagine you pick some dir but there are no git repositories, we probably should have some error UI

The existing syncer design streams out repos (and errors). You know when streaming is done (and therefore know if it is empty).

> I believe multiple root directories is the immediate need / very nice to have,

This is easy to support from a repo discovery / service connection perspective.

> (1) pass as command line arg

The PoC PR supports via envvar, but cli arg would be just as plausible. Keeping support for this as well as setup wizard from a backend perspective is straightforward.

> Also worth noting that we would like to allow modifying this local repo code host connection in sg admin portal

Right now the PoC uses the external service Other which already exists. This does not specify anything to do with were to find the local code. We could adjust the schema of the config and the serve-git API to support the usecase of setting the location of roots as a follow-up. Right now on startup we just always replace the external service for local code. IE a user can edit it, but the edits don't persist across restarts.

> We don’t actually define repos in the code host config, although I didn’t check what happens if we don’t use one of the hardcoded values and instead proceed to cloneURLs (here). Assuming we can’t utilize code host connection definition for adding/removing root directories I was thinking we just add a handler to app’s serve-git which updates Roots:

Indeed both are plausible. Changing our schema to support specifying root locations feels like the better solution since it is more visible / understandable.

> Looking at the PR, we have one serve-git and I was thinking we could have a lookup somewhere to map which directory in our list of “roots” will contain a given repo. Maybe this function can prepend the correct root by doing a lookup in the repo->root\_directory map, and it will then allow us to only spin up a single gitservice handler?

The URI field isn't visible to users and can encode this information. Alternatively the clone URL could encode absolute paths. We don't need the map.

> TLDR; If we don’t have a means of updating root directory/directories configuration via an external service configuration, then we can add a handler to serve git to allow these changes. Then update list repos handler to communicate to other services how they reach a given repo for sync. If it helps we can make a graphql query for frontend to call, which does the update.

We can add this. I think poking at serve-git via RPCs is not great since it doesn't solve the problem of where do we persist the location of roots across restarts.

> another question: The http.FileServer() seems like a nice to have and not necessary for any syncing.

Indeed, this is actually likely not necessary anymore. In the initial versions of src-expose we relied on git's "Dumb Protocol" for cloning. I can't remember when I added support for git's smarter protocol and why we kept the dumb protocol. If I had to guess I was worried about old versions of git in the src-cli users. But for app we don't need it, so will remove it.

> One more thought: for persisting desired local repo directories after startup, we might need to update the Other source and schema to take directories list. I mentioned earlier that we’d not store this in the ext svc configuration, but I forgot that we’d like to persist the list

Hah we came to the same thought as you can see in my comments above.

**Comment by vovakulikov** [comment]
@keegancsmith thanks for detailed reply. When you got a minute can you please break down in query/mutations examples how CRUD would work around this specific Other external service, I assume that there is nothing extraordinary since it's just special type of external service but I think at least configuration schema there is different. 

So 
- Do we need to create this external services or it will be pre-created automatically (when user starts their instance with local repo path through cli/terminal?)
- Can we have more than one "Other" external service? 
- How to update it if we want to change the local repositories path?

**Comment by gl-srgr** [comment]
@vovakulikov From my current work on this my understanding is the following:

> Do we need to create this external services or it will be pre-created automatically (when user starts their instance with local repo path through cli/terminal?)

Right now it'll be pre-created but this could change. The one scenario where we could choose to not create is if we let the user "disable" local repo sync when running sourcegraph app (by passing empty SRC env var, or something similar). If we want to create this Other ext svc during setup wizard, we need to read this [env var](https://github.com/sourcegraph/sourcegraph/pull/47673/files#diff-92222b9c3e2ceee22187e6eea95cd357c4cbfcf457a7350e86c38e8235b13c51R30) so that we can populate url in the config. We can use the same default value as well.

> Can we have more than one "Other" external service?

Yes, we could create multiple with all pointing to the same url or different, but right now the feature PR would just sync the same directory for all of the external services. With some additional changes we can have each Other external service define their own specific root directories.

> How to update it if we want to change the local repositories path?

This can be handled with some updates to external service config. I just wrapped up a draft PR to submit to @keegancsmith for feedback

**Comment by keegancsmith** [comment]
@gl-srgr responses look correct to me and his draft PR looks good. We can make this work however we want, but I think the clean thing for now is that the wizard creates/mutates the other external service, just like it will for all code hosts. The schema is in the file `other_external_service.schema.json`.

I'll take a closer look later today at the draft PR and we can probably land it soon. FYI the mechanics of how this all works can change. I imagine with the wizard we will likely get rid of the SRC envvar I added and do something which is more convenient to the user. Using the envvar is just convenient while we play around with code.

**Comment by gl-srgr** [comment]
@slimsag I tagged you in the draft PR but I didn't provide much context; the recent comments here have more context than the PR.

There was a meeting a few weeks ago about which user story to accomodate:

1. Skip setup wizard: Less steps, less guidance, so quicker to reach value. However, the user needs to be more familiar with sourcegraph.
2. Provide setup wizard: More steps; seems like a lot of steps altogether since we have image setup steps before sg starts. However, the user gets guidance if they are not familiar with sg.

The SRC env var can be used for (1) which @slimsag was leaning towards. However, this means in the background we have two flows for managing local repo sync directories.

We can make this decision in a separate PR from my current one. Just wanted to mention these two different user stories

**Comment by keegancsmith** [comment]
FYI I am closing this issue out since we do have local code and this is listed as a must have for starship. If our implementation is lacking something please file a specific follow-up issue.

There is a lot of great context/discussion above which will be useful to revisit when we want to improve the interactions with local code.
